server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -571,13 +580,286 @@ protected void doRun() {
+            this.slot = slot;
+            this.indexRequest = indexRequest;
+            this.pipelines = pipelines;
+            this.hasFinalPipeline = hasFinalPipeline;
+        }
+
+        public int getSlot() {
+            return slot;
+        }
+
+        public IndexRequest getIndexRequest() {
+            return indexRequest;
+        }
+
+        public List<String> getPipelines() {
+            return pipelines;
+        }
+
+        public boolean isHasFinalPipeline() {
+            return hasFinalPipeline;
+        }
+    }
+
+    private void executePipelinesInBatchRequests(
+        final List<Integer> slots,
+        final Iterator<String> it,
+        final boolean hasFinalPipeline,
+        final List<IndexRequest> indexRequests,
+        final IntConsumer onDropped,
+        final BiConsumer<Integer, Exception> onFailure,
+        final AtomicInteger counter,
+        final BiConsumer<Thread, Exception> onCompletion,
+        final Thread originalThread
+    ) {
+        if (indexRequests.size() == 1) {
+            executePipelines(
+                slots.get(0),
+                it,
+                hasFinalPipeline,
+                indexRequests.get(0),
+                onDropped,
+                onFailure,
+                counter,
+                onCompletion,
+                originalThread
+            );
+            return;
+        }
+        while (it.hasNext()) {
+            final String pipelineId = it.next();
+            try {
+                PipelineHolder holder = pipelines.get(pipelineId);
+                if (holder == null) {
+                    throw new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist");
+                }
+                Pipeline pipeline = holder.pipeline;
+                String originalIndex = indexRequests.get(0).indices()[0];
+                Map<Integer, IndexRequest> slotIndexRequestMap = createSlotIndexRequestMap(slots, indexRequests);
+                innerBatchExecute(slots, indexRequests, pipeline, onDropped, results -> {
+                    for (int i = 0; i < results.size(); ++i) {
+                        if (results.get(i).getException() != null) {
+                            IndexRequest indexRequest = slotIndexRequestMap.get(results.get(i).getSlot());
+                            logger.debug(
+                                () -> new ParameterizedMessage(
+                                    "failed to execute pipeline [{}] for document [{}/{}]",
+                                    pipelineId,
+                                    indexRequest.index(),
+                                    indexRequest.id()
+                                ),
+                                results.get(i).getException()
+                            );
+                            onFailure.accept(slots.get(i), results.get(i).getException());
+                        }
+                    }
+
+                    Iterator<String> newIt = it;
```

Comment:
I'm not clear what this does?

Can we have either a comment or a more self-documenting variable name?